Source/ASDisplayNode+InterfaceState.m

```diff
@@ -0,0 +1,154 @@
+//
+//  ASDisplayNode+InterfaceState.m
+//  Texture
+//
+//  Copyright (c) 2017-present, Pinterest, Inc.  All rights reserved.
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+
+#import <AsyncDisplayKit/ASDisplayNode+InterfaceState.h>
+
+@interface ASDisplayNodeInterfaceDelegateManager ()
+{
+    NSHashTable *_interfaceDidChangeDelegates;
+    NSHashTable *_interfaceDidEnterVisibleDelegates;
+    NSHashTable *_interfaceDidExitVisibleDelegates;
+    NSHashTable *_interfaceDidEnterDisplayDelegates;
+    NSHashTable *_interfaceDidExitDisplayDelegates;
+    NSHashTable *_interfaceDidEnterPreloadDelegates;
+    NSHashTable *_interfaceDidExitPreloadDelegates;
+    NSHashTable *_interfaceNodeDidLayoutDelegates;
+    NSHashTable *_interfaceNodeDidLoadDelegates;
+}
+@end
+
+@implementation ASDisplayNodeInterfaceDelegateManager
+
+- (instancetype)init
+{
+    if (self = [super init]) {
+        _interfaceDidChangeDelegates = [NSHashTable weakObjectsHashTable];
+        _interfaceDidEnterVisibleDelegates = [NSHashTable weakObjectsHashTable];
+        _interfaceDidExitVisibleDelegates = [NSHashTable weakObjectsHashTable];
+        _interfaceDidEnterDisplayDelegates = [NSHashTable weakObjectsHashTable];
+        _interfaceDidExitDisplayDelegates = [NSHashTable weakObjectsHashTable];
+        _interfaceDidEnterPreloadDelegates = [NSHashTable weakObjectsHashTable];
+        _interfaceDidExitPreloadDelegates = [NSHashTable weakObjectsHashTable];
+        _interfaceNodeDidLayoutDelegates = [NSHashTable weakObjectsHashTable];
+        _interfaceNodeDidLoadDelegates = [NSHashTable weakObjectsHashTable];
```

Comment:
On second thought, using `-respondsToSelector:` is only a tiny bit slower and allows for proxying, so that's probably the way to go. Still fast.

Reply:
I don't think we should have to create all these up front...since we would only have to use the respondsToSelector: once per class (not even once per delegate), we should get a significant win by relying on just one object tracking the delegates, and we could use a simple bitfield / NS_OPTIONS to cache the ASInterfaceStateDelegateMethodsImplemented (similar to the flags in ASCollection that cache what the delegate / datasource implement, but just an NS_OPTIONS).

Reply:
In recent versions of Objective-C, `respondsToSelector:` is truly very fast. It amortizes to 1 additional objc_msgSend, 3 static function invocations, and one inline-assembly routine. It also has the added benefit of checking-from and storing-to the ObjC impcache, so the subsequent message-send is optimized.

Writing our own bitfield-by-class would amount to us reimplementing that portion of the impcache, without the benefit of direct access to the class method table. And even accessing our table – say it were in a singleton registry object – would require retaining-releasing it unless we coded around that, which would easily subsume the entire rest of the lookup. The added boilerplate is also nontrivial.

So there is a win available here, but I believe it's small. Let's avoid these calls only if we can do it in a simple and tremendously fast & scalable way. If we can't, then we shouldn't worry about it – I believe we can absolutely achieve far superior optimizations by using resources for other parts of the framework (such as transfer-collections or more judicious applications of `__unsafe_unretained`). 

Implementations;
- [`-respondsToSelector:`](https://opensource.apple.com/source/objc4/objc4-723/runtime/NSObject.mm)
- [`class_respondsToSelector_inst`](https://opensource.apple.com/source/objc4/objc4-723/runtime/objc-class.mm.auto.html)
- [`lookupImpOrForward` and `lookupImpOrNil`](https://opensource.apple.com/source/objc4/objc4-723/runtime/objc-runtime-new.mm)
- [ARM64: `cache_getImp`, used by both `objc_msgSend` and `lookupImpOrNil`](https://opensource.apple.com/source/objc4/objc4-723/runtime/Messengers.subproj/objc-msg-arm64.s)